### ISSUE TITLE
Add "ray_thickness" property to Raycast2D

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#2577]
### Describe the project you are working on
Projects with projectiles, shooters, platformers ect

### Describe the problem or limitation you are having in your project
Using multiple rays for when you really want a single thick ray is both a hassle and unreliable, because inaccurate. Whatever you do, there are always gaps between your rays.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
One ray instead of many rays with gaps would mean much easier, much more intuitive and way more reliable work with raycasts.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
![image](https://user-images.githubusercontent.com/47016402/114171744-0324ec80-9935-11eb-9b77-a71c41db9817.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?
One raycast is rarely ever enough, so I would say this is used often and to my knowledge can't be worked around.

### Is there a reason why this should be core and not an add-on in the asset library?
It would improve user friendlyness and reliability of an existing feature.

**Comment by Xrayez** [comment]
I think that ray casting cannot have a notion of width, only length. It's like segment intersection on mathematical level. If this proposal is implemented as-is, then it would require rewriting the underlying physics backends on a deeper level.

However, what you ask for can be achieved with *shape* casting, see #710. You'd need to assign a `RectangleShape2D` and align it to work like ray cast:

![image](https://user-images.githubusercontent.com/17108460/114175236-73863a80-9942-11eb-8585-5dd3eda38656.png)

This feature is available in [Goost](https://github.com/goostengine/goost), by the way.

**Comment by pycbouh** [comment]
In addition to shapecasting, you can just adjust your rays. Instead of adding a single one at the center, add two at the edges. Add more rays in the middle to have wider coverage. Those are not that computationally hard to use them so sparsely.

**Comment by golddotasksquestions** [comment]
@pycbouh Thanks, but I'm well aware about this practice. I thought I was very clear about it when I wrote "Using multiple rays ..." right in the beginning.

I submitted this proposal issue because I am really unsatisfied with using multiple rays (which I do often). Searching for Shape casting alternatives only led me to Godot's [Raycast Shapes](https://godotengine.org/article/godot-31-will-get-many-improvements-kinematicbody) which has a bit unfortunate name as it sounds so much alike, but has nothing to do with this issue at all. I must have missed [Xrayez proposal](https://github.com/godotengine/godot-proposals/issues/710) though. 
Still, casting a shape is not quite the same as the intuitiveness and simplicity that would come with a ray that has a thickness property, but I also understand that mathematically the rays used for raycasting don't have any thickness and changing that would might be difficult and costly in devtime if possible/reasonable at all.
I could not test ShapeCasting as Godot does not have it build in in the High Level API and I hear that there is a low level approach for ShapeCasting for the first time now.  

I thought I still put this up here to express my desire just in case someone _does_ see a efficient way to implement it and feels inspired.

**Comment by golddotasksquestions** [comment]
> Gaps between the rays are fine, because there is really no perfect physics in games, it's all an illusion.

In my experience gaps are sometimes fine, often they are not. I don't write issues because I am board, I write them after being frustrated and annoyed with something I have used a lot, and tried what I can to find workarounds. Maybe for you in all your projects gaps are fine, and that is good.

**Comment by pycbouh** [comment]
I'm not sure why you choose to use such an agressive writing tone, but emitting a multitude of rays is a normal practice in gamedev. It's used everywhere by lots of developers. The only trick is to tinker the shape in which you cast your rays. That's all I'm saying here.

**Comment by golddotasksquestions** [comment]
There is no absolutely no aggressive mindset, intention and definitely no undertone whatsoever towards you, in my post or in my comments. Please don't read into things. I assume you having the best intentions, regardless what "tone" I might be inclined to read into your comments, please do the same for me.

**Comment by elvisish** [comment]
Is shape casting possible yet in node form? It’s so easy to add a raycast node and set it up visually with a node (something Unity never got) and it would be great to do the same with a shape.

Also, multiple casts requires looping through each raycast node for collision checks which is surely not as performant as a shape.

**Comment by schneediener** [comment]
> @pycbouh There is no absolutely no aggressive mindset, intention and definitely no undertone whatsoever towards you, in my post or in my comments. Please try not to read into things. I assume you are having the best and only positive intentions, regardless what "tone" I might be inclined to read into your comments, please do the same for me.

Unlikely to still care about this, but I believe we can achieve some version of this by adjusting the 'scale' in only the respective width axis. Appears to increase the thickness of the raycast, from my testing.

**Comment by elvisish** [comment]
Currently doing raycasting in godot 4 with:
```
var query = PhysicsRayQueryParameters2D.create(from, to)
var collision = get_world_2d().direct_space_state.intersect_ray(query)
```
A very simple way of getting raycast results, but can shapecasts be used like this? It would still be nice to be able to define a thickness to an intersect_ray.

**Comment by PerMalmberg** [comment]
> Currently doing raycasting in godot 4 with:
> 
> ```
> var query = PhysicsRayQueryParameters2D.create(from, to)
> var collision = get_world_2d().direct_space_state.intersect_ray(query)
> ```
> 
> A very simple way of getting raycast results, but can shapecasts be used like this? It would still be nice to be able to define a thickness to an intersect_ray.

Came here, searching for the very same thing.

**Comment by AThousandShips** [comment]
@PerMalmberg Please don't bump without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by elvisish** [comment]
> @PerMalmberg Please don't bump without contributing significant new information. Use the :+1: reaction button on the first post instead. 

Or you could encouragingly let @PerMalmberg know it has been implemented and possibly close this proposal:
https://docs.godotengine.org/en/3.6/classes/class_shapecast.html
https://docs.godotengine.org/en/stable/classes/class_shapecast3d.html

**Comment by AThousandShips** [comment]
Well I couldn't have known a random proposal that wasn't closed by someone else, no need for that attitude over a basic moderation message...

**Comment by Calinou** [comment]
I don't see how OP's proposal can be implemented without shapecasting, or throwing multiple rays that are close enough to each other. (How close they should be will depend on your needs, so this would mean adding *another* property to control ray separation.)

**Comment by AThousandShips** [comment]
A relevant difference is that `ShapeCast` can't be configured to have an arbitrary length, so this would be accomplished by adding a new type of shape cast that can be configured differently

**Comment by elvisish** [comment]
> no need for that attitude over a basic moderation message...

No attitude intended, I just think offering a suggestion to the poster would be more useful and engaging than telling them off for bumping (if they're totally new to the proposals, it could put them off completely)

And yeah I did comment two and half years ago, before shapecasting was added?

**Comment by AThousandShips** [comment]
Fair enough, but informing people not to bump is standard procedure, and it's impossible to be on top of *every* proposal and issue across the codebase, and I have other things to focus on than going through things for each time someone bumps something

So I'd suggest that next time, instead of correcting someone for doing moderation, just give the suggestion yourself

**Comment by elvisish** [comment]
> Fair enough, but informing people not to bump is standard procedure

It can be done politely, and the standard no bump message is very dismissive and discouraging to people, especially to new Godot users (and if it's a very old proposal that no-ones seen in years they might not have any new information to add).

> So I'd suggest that next time, instead of correcting someone for doing moderation, just give the suggestion yourself

I did, I posted the links myself on my reply.

**Comment by AThousandShips** [comment]
Well thank you for your feedback, but there's not much else to say than that, discouraging bumping isn't for no reason

I'll go ahead and mark this conversation as off topic, have a nice day
